arch-split Refine up to Invariants_H for AARCH64 (Draft)

Xaphiosis opened the pull request:
🦆🦆🦆 This PR needs feedback before I deploy anything to the other arches or move further. In an attempt to make the diff a bit more useful for reviewing than "new files added with stuff", `Invariants_H` got duplicated into each of the new files, and then clobbered by the actual file contents.
**Do not attempt to review the whole thing at once**, or the "duplicate and move Invariants_H" commit which will be squashed into once all this works.
The Refine fixup commit is not really indicative of any policies I'm aiming for in the future, it's just there so Refine builds.
The plan is: once we are OK with the way things are done, I'll expand this PR with deployment to other arches, fix up CRefine, and then we can merge it to master before continuing the arch-split train.

Invariants_H gets split into 4 files processed in this order:
    - InvariantsPre_H: initial setup
    - ArchInvsDefs_H: arch-specific invariant property definitions
    - Invariants_H: remains to store arch-generic
      invariants/properties/lemmas
    - ArchInvsLemmas: locale interpretations and arch-specific
      invariant-related lemmas

Several lemmas/interpretations/locales got moved to InvariantUpdates_H in order to resolve dependencies without introducing more files.

AARCH64 Refine will build, but CRefine is currently broken.

corlewis commented on proof/refine/InvariantsPre_H.thy:
```diff
@@ -5,45 +5,76 @@
 
-declare lookupPTSlotFromLevel.simps[simp del]
-declare lookupPTFromLevel.simps[simp del]
+section \<open>Locale Setup for Idempotence over kernel_state Field Updates\<close>
```
Very nice job collecting and rationalising these locales!

Just trying to think of what might help us in the future, do we want a comment here with a quick explanation of what we mean by idempotence over field updates, and maybe with an explanation that the locales are actually populated and interpreted in Invariants_H? It would be harder to explain here, but it might also help to have a comment explaining that completing the lattice is helpful to avoid future 'landmines', even if some of the locales aren't used now.

More ambitious ideas that almost definitely aren't worth the time would be to have some sort of test that confirms that the lattice is complete, and to have a new tool that constructs the lattice given `Arch` and the base locales.

Xaphiosis replied:
comments added

lsf37 replied:
I'm happy with the new comments.

corlewis replied:
Same

corlewis commented on proof/refine/AARCH64/ArchInvsDefs_H.thy:
```diff
@@ -1,35 +1,32 @@
 (*
- * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2024, Proofcraft Pty Ltd
```
I thought we want to keep the earlier dates? If I'm wrong with that, you could also change the copyright in `InvariantsPre_H`.

lsf37 replied:
This is a new file, it just looks like it changed, because Raf set it up so there is a more useful diff. I guess it is debatable whether the content is really new if it's mostly moved, but there is also enough new stuff in here to actually call this new material from scratch.

corlewis replied:
Hmm, I guess there's trade-offs. If we're thinking of it as new content should it keep the Data61 copyright?

Xaphiosis replied:
I believe so, I'm taking other people's stuff, rearranging it in a different way, and then adding my own stuff. So that bumps the Proofcraft copyright and keeps the one that made the old stuff.

corlewis replied:
For my clarity, in general if a change is 'just' updating existing content then we don't want to bump the copyright, right? I know that there's a spectrum between updating and new content, but that was my understanding and what led to this comment initially.

corlewis replied:
On the topic, how about `Invariants_H` in this PR? In my mind that one at least should keep the original copyright.

Xaphiosis replied:
It's hard to know what the threshold is. We're adding requalifies, rearranging stuff with the locales, renaming the locales ... @lsf37 ?

corlewis replied:
My understanding is that how large a change to existing content is is relevant to whether we add the Proofcraft copyright or not, but that once the copyright is there we want to keep the original date. I might be thinking about this wrong of course, but I actually can't think of a situation where we want to bump a copyright. We could maybe make it a range and say something like `Copyright 2023-2024` but I don't know if that actually benefits us in anyway.

corlewis replied:
Now that I've overthought this so much, I'm actually back to questioning the original copyright I commented on, in `ArchInvsDefs_H.thy`. We're either treating it as new content, in which case it shouldn't have the Data61 copyright, or we're considering it as based on the old content from `Invariants_H.thy` and it should keep the existing copyright.

Xaphiosis replied:
copyrights reverted

lsf37 commented on proof/refine/AARCH64/ArchInvsDefs_H.thy:
```diff
@@ -1,35 +1,32 @@
 
-theory Invariants_H
+(* This theory contains the arch-dependent part of invariant definitions.
+   Due to the large amount of definitins, we opt for a Pre file rather than fixing constants in
+   locales. *)
```
definitins -> definitions

Xaphiosis replied:
thanks

lsf37 commented on proof/refine/AARCH64/ArchInvsDefs_H.thy:
```diff
@@ -39,123 +36,20 @@ end
-
-abbreviation sym_heap_sched_pointers :: "global.kernel_state \<Rightarrow> bool" where
-  "sym_heap_sched_pointers s \<equiv> sym_heap (tcbSchedNexts_of s) (tcbSchedPrevs_of s)"
+(* FIXME arch-split: currently global in Invariants_H
+definition max_ipc_words :: machine_word where
+  "max_ipc_words \<equiv> capTransferDataSize + msgMaxLength + msgMaxExtraCaps + 2" *)
```
I think this definition is the same on all architectures, so it is correct for it to be global. The ingredients may be arch specific, but how the total size is computed is generic.

Xaphiosis replied:
removed

lsf37 commented on proof/refine/AARCH64/ArchInvsDefs_H.thy:
```diff
@@ -39,123 +36,20 @@ end
+  "max_ipc_words \<equiv> capTransferDataSize + msgMaxLength + msgMaxExtraCaps + 2" *)
 
 definition tcb_cte_cases :: "machine_word \<rightharpoonup> ((tcb \<Rightarrow> cte) \<times> ((cte \<Rightarrow> cte) \<Rightarrow> tcb \<Rightarrow> tcb))" where
  "tcb_cte_cases \<equiv> [ 0 << cteSizeBits \<mapsto> (tcbCTable, tcbCTable_update),
```
I would have expected this one to be a generic definition. Do we we have architectures that add TCB cap slots?

Xaphiosis replied:
No, but other than AARCH64 and RISCV64, they do this:
```
184: "tcb_cte_cases \<equiv> [ 0 \<mapsto> (tcbCTable, tcbCTable_update),
185-                   16 \<mapsto> (tcbVTable, tcbVTable_update),
186-                   32 \<mapsto> (tcbReply, tcbReply_update),
187-                   48 \<mapsto> (tcbCaller, tcbCaller_update),
188-                   64 \<mapsto> (tcbIPCBufferFrame, tcbIPCBufferFrame_update) ]"
```

lsf37 commented on proof/refine/InvariantsPre_H.thy:
```diff
@@ -5,45 +5,76 @@
 
-declare lookupPTSlotFromLevel.simps[simp del]
-declare lookupPTFromLevel.simps[simp del]
+section \<open>Locale Setup for Idempotence over kernel_state Field Updates\<close>
```
Part of the confusion in the comment may be that "idempotence" is not the right word. Idempotence means `f (f x) = f x`, i.e. applying more than once is the same as applying once. What we are proving here is more identity, i.e. f x = x, but with side conditions. The locales are identifying which updates are the identity function on specific predicates. For instance, when a function doesn't change the `ksPSpace` component of the state, that function is the identity on all predicates that only depend on `ksPSpace`.

+1 on adding where these are populated and interpreted, I think that would be a good idea.

The setup is good, though, I'm happy with making this whole setup a bit more principled.

Xaphiosis replied:
> +1 on adding where these are populated and interpreted, I think that would be a good idea.

What do you mean by this?

corlewis replied:
I assume that that was referring to my earlier [comment](https://github.com/seL4/l4v/pull/833#discussion_r1872729019), where I suggested "maybe with an explanation that the locales are actually populated and interpreted in Invariants_H".

lsf37 replied:
yeah, what Corey said. I didn't realise my comment would not show up under Corey's comment (it did in the review view, but apparently not here).

Xaphiosis replied:
Hopefully that's done, let me know if it explains sufficiently.

lsf37 replied:
The section title still has "Idempotence" in it

lsf37 replied:
Apologies, it does not, I was staring at the old version. Happy with the new version!

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -223,21 +174,7 @@ primrec live0' :: "Structures_H.kernel_object \<Rightarrow> bool" where
-   | AARCH64_H.KOPTE pte \<Rightarrow> {}
-   | AARCH64_H.KOVCPU vcpu \<Rightarrow> vcpu_tcb_refs' (AARCH64_H.vcpuTCBPtr vcpu)"
-
+(* FIXME arch-split: do we want to expose this, or put it in Arch? *)
 definition hyp_refs_of' :: "kernel_object \<Rightarrow> (obj_ref \<times> reftype) set" where
```
I think we're exposing it AInvs as well, so I'd keep it set up the same way as there.

Xaphiosis replied:
ok will remove FIXME

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -536,27 +369,6 @@ definition valid_ntfn' :: "Structures_H.notification \<Rightarrow> kernel_state
   | ActiveNtfn b \<Rightarrow> True)
   \<and> valid_bound_tcb' (ntfnBoundTCB ntfn) s"
 
-(* FIXME AARCH64: remove if unused at the end *)
-definition valid_mapping' :: "machine_word \<Rightarrow> vmpage_size \<Rightarrow> kernel_state \<Rightarrow> bool" where
```
Thanks for removing it!

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -1317,12 +1081,11 @@ lemma objBitsKO_Data:
-lemmas untypedBits_defs = minUntypedSizeBits_def maxUntypedSizeBits_def
-lemmas objBits_simps = objBits_def objBitsKO_def word_size_def archObjSize_def
-lemmas objBits_simps' = objBits_simps objBits_defs
+(* FIXME arch-split: why are these visible outside of Arch?! *)
+thm tcbBlockSizeBits_def epSizeBits_def word_size_def
```
I guess they always have been, because it was too much work. If we accept `word_size_def` as generic, then `epSizeBits_def` would be as well, but `tcbBlockSizeBits_def` should probably not be. I guess one thing that is happening is that this is not properly arch split in C either: additional TCB fields are just `ifdef`ed instead of actual arch fields. Doesn't mean we can't properly split it on the formalisation side, of course.

Xaphiosis replied:
I removed this FIXME but still think that we should examine hiding tcbBlockSizeBits. Maybe in a separate PR as that'll change the spec and probably break things.

lsf37 replied:
Happy with that.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -1289,14 +1055,12 @@ lemma ntfn_splits[split]:
 
 section "Lemmas"
 
-schematic_goal wordBits_def': "wordBits = numeral ?n" (* arch-specific consequence *)
-  by (simp add: wordBits_def word_size)
```
If we leave `word_size_def` generic, we can probably leave this one generic as well.

Xaphiosis replied:
From MM discussion: I will try keep it internal as an extra guard rail, even though word_size itself is exported. If things get too bad, we can abort the guard rail.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -1390,27 +1150,10 @@ lemma obj_atI' [intro?]:
   by (simp add: cte_wp_at'_def)
 
-
-lemma tcb_cte_cases_simps[simp]:
-  "tcb_cte_cases 0  = Some (tcbCTable, tcbCTable_update)"
-  "tcb_cte_cases 32 = Some (tcbVTable, tcbVTable_update)"
-  "tcb_cte_cases 64 = Some (tcbReply, tcbReply_update)"
-  "tcb_cte_cases 96 = Some (tcbCaller, tcbCaller_update)"
-  "tcb_cte_cases 128 = Some (tcbIPCBufferFrame, tcbIPCBufferFrame_update)"
```
With the numbers unfolded, these would be arch specific (or at least mode/machine word size specific). If we could avoid unfolding `cteLevelBits` in the proofs, then the lemma would stay generic or would not even be needed. Not sure it's worth the hassle at this point, but might be worth a FIXME.

lsf37 replied:
Thinking more about it, the main reason why we tend to unfold these is because it is automatic for Isabelle to decide that e.g. `0 != 32`, but it can't automatically decide that `0 * cteLevelBits != 1 * cteLevelBits`. There is a mechanism somewhere for turning a concrete distinct list into a set of inequality `simp` lemmas. We could use that here, and prove `distinct [0 * cteLevelBits, 1 * cteLevelBits, ...]` once and use that. Then Isabelle should be able to do symbolically what it now needs the numbers for. I'll try find what that tool is called.

lsf37 replied:
Right, it's not a tool, and it's not merged yet, it's just an application of `simplified`. Here is an example: https://github.com/seL4/l4v/blob/dc2fd26954724fca79d175528726314de834605a/proof/crefine/ARM/CSpace_C.thy#L3064-L3088

Xaphiosis replied:
I've deployed something, it'll need review once I push it

Xaphiosis replied:
Please see the eventual infrastructure and the FIXMEs about overlapping lemmas. It follows on from our discussion on MM about being unable to sort the conjuncts and remove the redundant ones, and I still don't know how to do that.

lsf37 replied:
:+1: Happy with that for now. To make it nice, we probably have to adjust some of these TCB/CTE proofs quite a bit, which may even reach into spec changes. Something for some time in the future, but not for the current arch split pass.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -1800,26 +1483,6 @@ lemma obj_at'_pspaceI:
   "obj_at' t ref s \<Longrightarrow> ksPSpace s = ksPSpace s' \<Longrightarrow>  obj_at' t ref s'"
   by (auto intro!: projectKO_stateI simp: obj_at'_def ps_clear_def)
 
-lemma cte_wp_at'_pspaceI:
-  "\<lbrakk>cte_wp_at' P p s; ksPSpace s = ksPSpace s'\<rbrakk> \<Longrightarrow> cte_wp_at' P p s'"
-  apply (clarsimp simp add: cte_wp_at'_def getObject_def)
```
I would have expected this one also to be a generic lemma. Does it need any arch specific ingredients?

Xaphiosis replied:
I was able to prove this with better tcb_cte_cases infrastructure. next round

Xaphiosis replied:
done

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -2148,109 +1778,23 @@ lemma tcb_ctes_clear:
-  using assms by (simp add: cte_wp_at_cases' cte_level_bits_def objBits_defs)
-
-lemma cte_wp_at_tcbI':
+(* FIXME arch-split: naming: do we want to use the opaque naming style of Invariants_AI_2? *)
+locale Invariants_H_cte_ats =
```
I like proper names better, at least if they make sense. Sometimes there is too much stuff in there for a good name, but `cte_ats` here is good.

lsf37 replied:
In this case, if we can make `tcb_cte_cases` generic, I suspect that this locale becomes unneccessary.

Xaphiosis replied:
I don't know, `cte_wp_at_cases'` is pretty nasty and I don't have the time available to go through it carefully... not seeing a sensible solution inside.

lsf37 replied:
No worries, fine to leave for another time when we have better infrastructure for it.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -2437,44 +1941,6 @@ lemma typ_at_lift_valid_untyped':
-  unfolding frame_at'_def
-  by (wpsimp wp: hoare_vcg_all_lift hoare_vcg_const_imp_lift assms split_del: if_split)
-
-lemma typ_at_lift_valid_cap':
-  assumes P: "\<And>P T p. \<lbrace>\<lambda>s. P (typ_at' T p s)\<rbrace> f \<lbrace>\<lambda>rv s. P (typ_at' T p s)\<rbrace>"
-  shows      "\<lbrace>\<lambda>s. valid_cap' cap s\<rbrace> f \<lbrace>\<lambda>rv s. valid_cap' cap s\<rbrace>"
```
Could this one be built on top of a new (similar) lemma about `valid_arch_cap'`?

lsf37 replied:
(not sure it's really worth it here, because it's year another arch thing that needs to be proved first, unless the entire `typ_at_lift_valid_cap` goes into a later theory)

Xaphiosis replied:
Yes, because it is a locale assumption, once it is satisfied it becomes available in general theory. I do not know whether we need a separate lifter for `valid_arch_cap`. My current guess is we shouldn't need one.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -2675,10 +2118,48 @@ lemma valid_mdb_ctesI [intro]:
+  done
+
+
+section "Idempotence over kernel_state field updates"
```
Idempotence -> Identity

Xaphiosis replied:
I'll go with this phrasing

Xaphiosis replied:
done

lsf37 commented on proof/refine/AARCH64/ArchInvsLemmas_H.thy:
```diff
@@ -2191,1200 +165,104 @@ lemma cte_wp_at_cases':
-lemma invs_valid_pspace' [elim!]:
-  "invs' s \<Longrightarrow> valid_pspace' s"
-  by (simp add: invs'_def valid_state'_def)
+lemma cte_wp_atE'[consumes 1, case_names CTE TCB]:
+  assumes cte: "cte_wp_at' P ptr s"
```
There is already a comment about sorting through this. My gut feeling is that if tcb_cte_cases can be kept generic, a lot of these can be generic.

Xaphiosis replied:
do not see a way of making them generic, looked through them again, nothing jumps out for migration, so removed the browsing FIXME

lsf37 commented on proof/refine/AARCH64/ArchInvsLemmas_H.thy:
```diff
@@ -2191,1200 +165,104 @@ lemma cte_wp_at_cases':
-         simp add: ct_in_state'_def tcb_in_cur_domain'_def ct_idle_or_in_cur_domain'_def
-                   ct_not_inQ_def)
-  done
+(* arch-specific consequences of idempotence over kernel_state field updates *)
```
identity

lsf37 commented on proof/refine/AARCH64/ArchInvsLemmas_H.thy:
```diff
@@ -3449,99 +327,171 @@ lemma range_cover_canonical_address:
-   apply assumption
-  apply (fastforce dest: distinct_obj_range'_not_subset
-                   simp: obj_range'_def)
+(* FIXME arch-split: do we want a gen_typ_at_lifts version? *)
```
Would not be useful in the proofs, but might be useful in terms of making sure that each arch has all generic objects (in case we later ever add a generic object). So in terms of organising things, probably yes, but can be a FIXME for now.

Xaphiosis replied:
I managed to create one and put it in InvariantsUpdates_H. next round

lsf37 commented on proof/refine/AARCH64/ArchInvsLemmas_H.thy:
```diff
@@ -3449,99 +327,171 @@ lemma range_cover_canonical_address:
+lemmas bit_simps' = pteBits_def asidHighBits_def asidPoolBits_def asid_low_bits_def
+                    asid_high_bits_def bit_simps
+
+(* FIXME arch-split: do we want to export the gen ones? *)
+lemma objBitsT_simps:
```
Yes, I think that would be useful in proofs about those specific generic object types.

Xaphiosis replied:
done, but check how I did it, it's a bit height-reduced

lsf37 replied:
I think that's fine. We'll have to remember to update the number in `gen_objBitsT_simps` when we add a generic object, but I think that's reasonable.

lsf37 commented on proof/refine/AARCH64/InvariantUpdates_H.thy:
```diff
@@ -5,11 +5,92 @@
+
 (* FIXME: use locales to shorten this work *)
 
+(* FIXME arch-split: MOVE, or try make generic *)
+context Arch begin arch_global_naming
+
+lemma invs'_gsCNodes_update[simp]:
```
This one might bite for arches where things like valid_machie_state have more ingredients, but happy to leave it like this until it does bite. Maybe we're lucky.

Xaphiosis replied:
InvariantUpdates_H is not yet arch split, so will leave for this PR and see what happens later.

lsf37 commented:
The setup overall looks good from my side. Nice job with the update locales! I have minor comments only. If we can keep the main cte cases lemmas generic (they should be if we can abstract the lemma over the word size), then I think we'll save ourselves some pain down the road in CSpace theories that should be mostly generic, but won't be if tcb_cte_case is not.

corlewis commented on proof/refine/AARCH64/ArchInvsDefs_H.thy:
```diff
@@ -39,123 +36,20 @@ end
 
-end
-
-record itcb' =
```
Huh, not relevant for this PR but I'm mildly surprised that we don't have a Refine equivalent to AInvs' `iarch_tcb`. In AInvs that lets us reuse the `pred_tcb_at` machinery for vcpus as well.

Xaphiosis replied:
Then no one did such a thing in rt either?

corlewis replied:
Nope, rt so far has only been for architectures without hypervisor support, so there's never been a need for `pred_tcb_at` of an `arch_tcb` property.

Xaphiosis replied:
This strikes fear into my heart

Xaphiosis commented:
I fixed up non-proof items (including reverting the copyright changes) and added an explanation for the locale hierarchy based partly on Gerwin's explanation. Now I'll be looking at the proof changes, which will take longer.

corlewis commented on proof/refine/InvariantsPre_H.thy:
```diff
@@ -0,0 +1,129 @@
+  extensions of that locale. For arch-split there is an extra complication, where the conclusions
+  might use arch-specific facts, thus needing to be in Arch.
+
+  This means that for every one the update locales, we have to add an additional Arch locale.
```
```suggestion
  This means that for every one of the update locales, we have to add an additional Arch locale.
```

Xaphiosis replied:
Thanks, fixed, will include in next round.

Xaphiosis commented:
I've attempted to address all the issues raised so far, responded to the comments and fixed up Refine so it builds again. I will try to fix up CRefine next while I'm waiting for round 2 of reviewing.

Xaphiosis commented:
AARCH64 CRefine should be fixed now. That took longer than expected.

lsf37 commented on proof/refine/Invariants_H.thy:
```diff
@@ -1800,8 +1485,70 @@ lemma obj_at'_pspaceI:
+lemmas tcbSlot_defs = tcbCTableSlot_def tcbVTableSlot_def tcbReplySlot_def tcbCallerSlot_def
+                      tcbIPCBufferSlot_def
+
+lemma tcb_cte_cases_distinct:
+  "distinct [
+     tcbCTableSlot << cteSizeBits :: machine_word,
+     tcbVTableSlot << cteSizeBits,
+     tcbReplySlot << cteSizeBits,
+     tcbCallerSlot << cteSizeBits,
+     tcbIPCBufferSlot << cteSizeBits]"
+  (* FIXME arch-split: warning due to tcb_cte_cases_neqs overlapping *)
+  by (simp add: tcbSlot_defs tcb_cte_cases_neqs_n)
```
Cool. In the distant future, this could potentially be the form that we do not unfold further in other proofs and that the tcb cases then use everywhere. With that, there would be at least some hope that it is symbolic enough to not reduce to the same terms in `tcb_cte_cases_neqs` and therefore get rid of the warning (and the additional simplification work Isabelle has to do in each proof).

Xaphiosis replied:
I'm not following by what you mean here. If you mean that we don't unfold the `tcbCTableSlot` etc so we end up only needing these versions, then that would be nice, and would indeed not cause duplication in the `_neqs` lemma, which does normalise things around `0` and `1`. Don't think `Distinct_Cmd` will help much with that, it will either give a result not in simp normal form, or if we use `[simplified]`, duplicates as before.

Xaphiosis replied:
I pushed a fix for this, which also removes the FIXMEs. Playing around with `Distinct_Cmd` didn't really improve over `distinct []` and your attribute hack, but I guess it in theory could work around the attribute hack if it wasn't so awkward to use.

lsf37 replied:
> If you mean that we don't unfold the tcbCTableSlot etc

Yup, that's what I meant. I don't know where exactly the unfoldings come in, but in old code it could already be at the spec level. In any case, that's for another time.

Nice work on removing the duplicates!

lsf37 commented:
I'm happy with the second round and with the proof changes.

corlewis commented on proof/crefine/Move_C.thy:
```diff
@@ -1119,45 +1133,54 @@ lemma updateObject_cte_tcb:
+               split: if_split_asm
+               cong: if_cong)
+
+lemma self_elim:
+  "\<And>P Q. \<lbrakk> P \<Longrightarrow> Q; P \<rbrakk> \<Longrightarrow> Q"
+  by blast
```
Out of curiosity, do you remember why this is needed?

Xaphiosis replied:
It's needed in the two lemmas below, where we have that Isabellism of
```
[| blah ==> thesis; asm1; asm2 |] ==> thesis
```
and this rule is what I came up with to unscrew that. If there's a nicer way, please let me know. This pattern comes up when using `obtain` which is from Isar. The HOL way, we'd get an existential.

Xaphiosis commented:
@corlewis I have squashed the commits resulting from PR discussion, so we're back to disjoint commits for your review:

1. clone Invariants_H
2. interesting arch-split updates to split Invariants_H
3. boring Refine updates to make Refine work
4. mostly boring updates to make CRefine work (some excitement in Move_C)

corlewis commented on proof/refine/Invariants_H.thy:
```diff
@@ -3115,48 +2642,16 @@ lemma sch_act_wf_cases:
-  apply (case_tac arch_kernel_object; simp add: archObjSize_def objBitsT_simps bit_simps')
-  done
-
-lemma typ_at_aligned':
-  "\<lbrakk> typ_at' tp p s \<rbrakk> \<Longrightarrow> is_aligned p (objBitsT tp)"
```
Is this no longer needed? It seems to have been removed completely.

Xaphiosis replied:
correct, it's not mentioned again in l4v/proofs

Xaphiosis commented on proof/refine/Invariants_H.thy:
```diff
@@ -3115,48 +2642,16 @@ lemma sch_act_wf_cases:
-  by (simp add: makeObject_simps objBits_simps bit_simps')+
-
-
-lemma objBitsT_koTypeOf :
-  "(objBitsT (koTypeOf ko)) = objBitsKO ko"
```
Good catch. Yes, that should become an interface lemma and be requalified, it's a generic consequence partly based arch-specific results, something we expect all arch stuff to conform to

Xaphiosis replied:
I will probably leave the `[simp]` for now, and re-examine when I get to the `declare`

corlewis commented on proof/refine/AARCH64/ArchInvsLemmas_H.thy:
```diff
@@ -5,1835 +5,36 @@
 
-lemma valid_cap'_pspaceI:
+lemma valid_cap'_pspaceI[Invariants_H_pspaceI_assms]:
   "s \<turnstile>' cap \<Longrightarrow> ksPSpace s = ksPSpace s' \<Longrightarrow> s' \<turnstile>' cap"
```
Probably not worth doing and definitely not as part of this PR, but some of these `pspaceI_assms` lemmas feel like they would have the same proof on all architectures and could be generic if the interface was a bit lower down.

Xaphiosis replied:
That's very possible, but it will require effort and I'm thinking it won't save all that many lines. We'll see if any other arch comes into conflict, I'll keep your comment in mind.

corlewis commented on proof/refine/AARCH64/InvariantUpdates_H.thy:
```diff
@@ -5,11 +5,102 @@
+  tcb_at_cte_at'
+  gen_objBitsT_simps
+
+(* arch-specific typ_at_lifts in Arch *)
+lemmas gen_typ_at_lifts =
+         typ_at_lift_tcb' typ_at_lift_ep' typ_at_lift_ntfn' typ_at_lift_cte' typ_at_lift_cte_at'
+         typ_at_lift_valid_untyped' typ_at_lift_valid_cap' valid_bound_tcb_lift
+         valid_arch_tcb_lift'
```
A note for the distant future, I think that this will be a problem in `rt` with the more complicated `typ_at_lifts` setup we have there. I think for that we might want to add locales and do the arch-split dance between `Invariants_H`  and `ArchInvsLemas_H`.

Probably not worth the effort, but in a perfect world it would be nice to pull that setup back to master to test out arch-splitting it there.

Xaphiosis replied:
I don't quite understand. What kind of setup does rt have? If it cheers you up, `gen_typ_at_lifts` is so far unused, it's only the consequence of hiding the arch-specific ones under `typ_at_lifts` in Arch, and I left this behind as a precaution.

Xaphiosis replied:
Given your comment and how it's unused, should I remove it to prevent a derp later?

corlewis replied:
In `rt` we have a couple of locales that contain all of the `typ_at'` lifting rules and that are then interpreted as needed for any function that preserves `typ_at'`. See the whole section building up to https://github.com/seL4/l4v/blob/rt/proof/refine/RISCV64/Invariants_H.thy#L2777 if you're particularly curious. When it comes time for it to be arch-split, I imagine we'll want to keep as much of that generic as possible to reduce duplication and risk of de-sync.

corlewis replied:
> Given your comment and how it's unused, should I remove it to prevent a derp later?

Hmm, sorry, it wasn't actually specifically `gen_typ_at_lifts` that I was commenting on, more the general approach with `typ_at_lifts` being in `Arch`. I just put this here as the place where some of it is made generic again.

Xaphiosis replied:
In that case, they'll end up more like the `_upd` locale hierarchy? I.e. locales get declared, we do what we can in generic, then add on the rest in Arch, and then (sigh) any stragglers that need the Arch ones in the next generic theory ... hmm. I wish there was a nicer way.

corlewis replied:
Right, comparing it to the `_upd` locales makes sense and that does give us an approach to follow when we get up to that. Hopefully it ends up being not too bad.

corlewis commented on proof/refine/AARCH64/InvariantUpdates_H.thy:
```diff
@@ -5,11 +5,102 @@
+imports ArchInvsLemmas_H
 begin
 
+(* generic consequences which require arch-specific proofs *)
```
Hmm, I understand that this is a convenient location due to the import graph, but these requalifies being put here does feel like a bit of a negative consequence of how we're doing this arch-split. They don't really have any connection with the rest of this file and feel pretty disconnected to `Invariants_H` and `ArchInvsLemmas_H`, which are the files that actually relate to them.

I don't have any better suggestions so am fine to leave them here for now, but I can imagine negative consequences going forwards, where we make changes or add new lemmas to `ArchInvsLemmas_H` and don't think to change here as well, resulting in a new, inconsistent approach.

Xaphiosis replied:
I agree it's not ideal. The issue is the second best place to put them is inside the Arch file, and then there's a risk of de-sync between arches (not that bad in general, but not fun when doing the arch-split itself).
Another way is to use locales, which is something that I'll be aiming for for the rest of the files, but for invariants the dependency graph was a bit much. Maybe I can add a comment to clarify or something?

corlewis replied:
Ah right, I hadn't thought of this being specifically an artefact of the complicated invariants files.

corlewis reviewed (commented):
This looks good to me as well and I just have minor comments or notes for the future. Thanks for also taking the opportunity to move things around to be more localised and have a better structure overall!
